api/dataset.go

```diff
@@ -382,6 +382,132 @@ func (api *DatasetAPI) addVersion(w http.ResponseWriter, r *http.Request) {
 	log.Debug("upsert version", log.Data{"dataset_id": datasetID, "edition": edition, "version": version})
 }
 
+func (api *DatasetAPI) putDataset(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	datasetID := vars["id"]
+
+	if r.Header.Get(internalToken) != api.internalToken {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	dataset, err := models.CreateDataset(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	if err := api.dataStore.Backend.UpdateDataset(datasetID, dataset); err != nil {
+		handleErrorType(err, w)
+		return
+	}
+
+	setJSONContentType(w)
+	w.WriteHeader(http.StatusOK)
+	log.Debug("update dataset", log.Data{"dataset_id": datasetID})
+}
+
+func (api *DatasetAPI) putVersion(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	datasetID := vars["id"]
+	editionID := vars["edition"]
+	versionID := vars["version"]
+
+	if r.Header.Get(internalToken) != api.internalToken {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	version, err := models.CreateVersion(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	currentVersion, err := api.dataStore.Backend.GetVersion(datasetID, editionID, versionID, "")
+	if err != nil {
+		handleErrorType(err, w)
+		return
+	}
+
+	// Check current state of version document;
+	// if published do not try to update document
+	if currentVersion.State == publishedState {
+		http.Error(w, fmt.Sprintf("Unable to update document, already published"), http.StatusForbidden)
+		return
+	}
+
+	// Combine update version document to existing version document
+	newVersion := createNewVersionDoc(currentVersion, version)
+
+	if err = models.ValidateVersion(newVersion); err != nil {
+		log.ErrorR(r, err, nil)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if err := api.dataStore.Backend.UpdateVersion(newVersion.ID, version); err != nil {
+		handleErrorType(err, w)
+		return
+	}
+
+	if version.State == publishedState {
+		if err := api.dataStore.Backend.UpdateEdition(newVersion.Links.Edition.ID, version.State); err != nil {
+			log.ErrorC("failed to update the state of edition document to published", err, nil)
+			handleErrorType(err, w)
+			return
+		}
+
+		// Pass in newVersion variable to include relevant data needed for update on dataset API (e.g. links)
+		if err := api.updateDataset(datasetID, newVersion); err != nil {
+			log.ErrorC("failed to update dataset document once version state changes to publish", err, nil)
+			handleErrorType(err, w)
+			return
+		}
+	}
+
+	if version.State == associatedState {
```

Comment:
Consider checking if the state has been changed from something to associated, instead of just checking it its associated. Any update while the state is set to associated will satisfy this condition.

Reply:
The reason for doing this is that someone may not update the state but they may update the collection_id at which point an update will need to occur to the dataset document (as long as the state is `associated`)

Reply:
👍